### DISCUSSION TITLE
Flagger Canary Stuck In Failed Status Forever

### DISCUSSION THREAD

**Discussion by imdelgado** [fluxcd/flux2#3816] [Q&A]
Hi all I'm having a problem with flagger/canary my environment is using spinnaker, istio and Prometheus to do the analysis, basically, i notice we're always getting a failed status when i check the Events on my canary resource i see:

`Events:
  Type     Reason  Age    From     Message
  ----     ------  ----   ----     -------
  Warning  Synced  97m    flagger  XXXX-v011-primary.XXXX not ready: waiting for rollout to finish: 0 out of 1 new replicas have been updated
  Warning  Synced  48m    flagger  XXXX-v012-primary.XXXX not ready: waiting for rollout to finish: 0 out of 2 new replicas have been updated
  Warning  Synced  5m30s  flagger  XXXX-v013-primary.XXXX not ready: waiting for rollout to finish: 0 out of 1 new replicas have been updated`
  
  I see that the traffic is correctly sent across both versions but the status is always Failed, is there a way to wait until my primary replica is ready?

**Comment by kingdonb** [comment]
It sounds like you're describing an issue when the canary has reached the threshold (say 50%) and it's time to upgrade the primary to match the canary. Does your deployment have health checks (readiness and/or liveness checks?) The important one here I think is the readiness check. If you don't have readiness checks, then Kubernetes will assume the pod is ready before it can ever serve requests, and the traffic will get routed there, active primary pods will be torn down etc. before the primary is ready to serve with new pods.

I'm not sure why this would cause your canary to hang and fail. Do you have the readiness checks?
